pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -148,46 +186,46 @@ func (r *LazyBinaryReader) Close() error {
 }
 
 // IndexVersion implements Reader.
-func (r *LazyBinaryReader) IndexVersion() (int, error) {
-	reader, wg, err := r.getOrLoadReader()
+func (r *LazyBinaryReader) IndexVersion(ctx context.Context) (int, error) {
+	reader, wg, err := r.getOrLoadReader(ctx)
 	if err != nil {
 		return 0, err
 	}
 	defer wg.Done()
```

Comment:
Opinionated ignorable nit: I would wrap the `Reader` returned by `getOrLoadReader` to:

```go
type lazyReader struct {
    Reader
    wg *sync.WaitGroup
}

func (lr lazyReader) release() {
    lr.wg.Done()
}
```

So here we could just do:
```suggestion
	reader, err := r.getOrLoadReader(ctx)
	if err != nil {
		return 0, err
	}
	defer reader.release()
```

Reply:
i totally agree. I just didn't want to extend the changeset further; i can do this after we merge